### PR TITLE
Gui: support auto grouping editing for PropertyEnumeration

### PR DESCRIPTION
**Pull request by realthunder**
Forum thread https://forum.freecadweb.org/viewtopic.php?f=22&t=65633&p=569085#p569085

### CODE REVIEW

**Comment by luzpaz** [comment]
CC stylesheet authors:  
@Chrismettal (behave) and @turn211 (ProDark)

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5513">pipeline status</a> for feature branch [PR_5513](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5513). Pipeline [464826177](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/464826177) was triggered at [345f070](https://github.com/FreeCAD/FreeCAD/pull/5513/commits/345f07080b58b31d999b80e8071bf19eddb399a3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5513">pipeline status</a> for feature branch [PR_5513](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5513). Pipeline [470441375](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/470441375) was triggered at [1c3becb](https://github.com/FreeCAD/FreeCAD/pull/5513/commits/1c3becb576d66267e0579d0d23440eaec164dd38). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by AjinkyaDahale** [comment]
@donovaly from the description sounds like a feature. Maybe tag "for 0.21"?

**Comment by donovaly** [comment]
Yes, it is a feature but was committed before the freeze. So it could be merged.

@realthunder , can you please rebase to fix the merge conflicts?

@chennes , do you have time to look at this PR?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5513">pipeline status</a> for feature branch [PR_5513](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5513). Pipeline [497808534](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/497808534) was triggered at [e276bd3](https://github.com/FreeCAD/FreeCAD/pull/5513/commits/e276bd3bcc82808e21817246283bc1d363b7fb60). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by donovaly** [comment]
@chennes , since #6401 was merged , could this go in now or should be postponed to FC 0.21? One commit seems to fix an issue therefore it could be considered according to our "rule".

File: src/Gui/propertyeditor/PropertyItem.cpp [hunk 1]
```diff
@@ -24,14 +24,18 @@
...
+# include <algorithm>
 # include <QComboBox>
 # include <QFontDatabase>
 # include <QLocale>
 # include <QPalette>
 # include <QPixmap>
 # include <QTextStream>
 # include <QTimer>
+# include <QApplication>
+# include <QPalette>
```
Comment (chennes) [inline src/Gui/propertyeditor/PropertyItem.cpp:36]: This is double-included -- @donovaly can you take a look at the includes here and see if it's just confused by the recent changes you've been making to the includes?

File: src/Gui/propertyeditor/PropertyItem.cpp [hunk 2]
```diff
@@ -2844,16 +2848,35 @@ void PropertyEnumItem::setValue(const QVariant& value)
...
+    QString text;
+    QString fullText;
+    std::shared_ptr<EnumItems> children;
+    EnumItem(const QString &t = QString(), const QString &f = QString())
+        :text(t), fullText(f)
+    {}
+    void populate(QMenu *menu);
+};
+
+class EnumItems : public std::vector<EnumItem>
```
Comment (chennes) [inline src/Gui/propertyeditor/PropertyItem.cpp:2863]: Is there a reason to leave this a full-blown class, and not just a `typedef`?

File: src/Gui/propertyeditor/PropertyItem.cpp [hunk 3]
```diff
@@ -2882,18 +2905,91 @@ void PropertyEnumItem::setEditorData(QWidget *editor, const QVariant& data) cons
...
+                enumItems->emplace_back(commonModes[i], mode);
+        }
+        auto children = enumItems;
+        int j = -1;
+        for (auto &field : fields) {
+            ++j;
+            field = field.trimmed();
+            auto it = children->end();
+            if (field.isEmpty()) {
+                if (children->size())
```
Comment (chennes) [inline src/Gui/propertyeditor/PropertyItem.cpp:2930]:
```suggestion
                if (!children->empty())
```

File: src/Gui/propertyeditor/PropertyItem.cpp [hunk 4]
```diff
@@ -2882,18 +2905,91 @@ void PropertyEnumItem::setEditorData(QWidget *editor, const QVariant& data) cons
...
+            children = it->children;
+        }
+    }
+
+    if (!enumItems) {
+        QComboBox *cb = new QComboBox(parent);
+        cb->setFrame(false);
+        cb->setDisabled(isReadOnly());
         cb->addItems(commonModes);
+        QObject::connect(cb, SIGNAL(activated(int)), receiver, method);
```
Comment (chennes) [inline src/Gui/propertyeditor/PropertyItem.cpp:2955]: Prefer new-style connection so we get compile-time checks.

File: src/Gui/propertyeditor/PropertyItem.cpp [hunk 5]
```diff
@@ -2882,18 +2905,91 @@ void PropertyEnumItem::setEditorData(QWidget *editor, const QVariant& data) cons
...
+        item.populate(menu);
+    button->setMenu(menu);
+    QObject::connect(menu, &QMenu::aboutToShow, this, [=]() {
+        menu->setMinimumWidth(button->width());
+    });
+    QObject::connect(menu, &QMenu::triggered, this, [=](QAction *action) {
+        button->setText(action->data().toString());
+        button->picked();
+    });
+    QObject::connect(button, SIGNAL(picked()), receiver, method);
```
Comment (chennes) [inline src/Gui/propertyeditor/PropertyItem.cpp:2972]: Prefer new-style connection so we get compile-time checks.

File: src/Gui/propertyeditor/PropertyItem.cpp (hunk 1 above)
Comment (donovaly) [reply src/Gui/propertyeditor/PropertyItem.cpp:36]: `# include <QPalette>`

Is already in the file, so this should not be added again.

`# include <QApplication>`

should only be included if necessary.

besides this, please add includes sorted. This way doubled entries can be avoided.

Comment (donovaly) [reply src/Gui/propertyeditor/PropertyItem.cpp]: Thanks, this is now resolved.

File: src/Gui/propertyeditor/PropertyItem.cpp (hunk 2 above)
Comment (realthunder) [reply src/Gui/propertyeditor/PropertyItem.cpp:2864]: Because it needs to be forward declared for use by `EnumItem::children`.

**Comment by donovaly** [comment]
Can you please squash the changes to one commit? This makes it easier to review.

**Comment by realthunder** [comment]
You mean the last two commits? I thought splitting makes better explanation?

File: src/Gui/propertyeditor/PropertyItem.cpp (hunk 4 above)
Comment (realthunder) [reply src/Gui/propertyeditor/PropertyItem.cpp:2958]: It is intended to connect to the signal handler pass in to `PropertyItem::createEditor()`. So before this API got updated to new-style connection, that's all we can do.

File: src/Gui/propertyeditor/PropertyItem.cpp (hunk 5 above)
Comment (realthunder) [reply src/Gui/propertyeditor/PropertyItem.cpp:2975]: Same as above

**Review by chennes — APPROVED** [review]
IMO this is fine to go in, if @donovaly is OK to merge it at this point in time. I tested the modification on Windows and it appears to work well, and the code is fine (as mentioned above, it's actually a couple years old, so uses the older style Qt connection syntax, but I don't see that as a dealbreaker).

**Comment by donovaly** [comment]
> IMO this is fine to go in

Very good. Merging then as really last PR that is not a pure bugfix.

@realthunder and @chennes , please help me to document this in the Wiki. You don't need to format it well, I can do this. It would help me to have the info written there then send me a message where you documented it and I will take care with our fellow Wiki guys to polish it.

**Comment by realthunder** [comment]
@donovaly The functionality is described in the forum post linked in OP. I have never edited FreeCAD wiki before. It seems I need to request for an account for that? Also, not sure where to put this entry. I think it should be somewhere in power user hub, but not sure exactly where.

**Comment by chennes** [comment]
I can write up the wiki entry when I am back at the office in a couple hours.

**Comment by chennes** [comment]
I put the documentation here: https://wiki.freecadweb.org/FeaturePython_Custom_Properties#App::PropertyEnumeration

**Comment by donovaly** [comment]
> I put the documentation here: https://wiki.freecadweb.org/FeaturePython_Custom_Properties#App::PropertyEnumeration

Thanks! Can you please additionally mention the new feature in the release notes: https://wiki.freecadweb.org/Release_notes_0.20

**Pull request merged**